pkg/gofr/grpc.go

```diff
@@ -39,8 +43,38 @@ func (g *grpcServer) Run(c *container.Container) {
 		return
 	}
 
-	if err := g.server.Serve(listener); err != nil {
-		c.Logger.Errorf("error in starting gRPC server at %s: %s", addr, err)
-		return
+	go func() {
+		if err := g.server.Serve(listener); err != nil {
+			c.Logger.Errorf("error in starting gRPC server at %s: %s", addr, err)
+		}
+	}()
+
+	// Handle system signals for graceful shutdown
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM)
+
+	go func() {
+		s := <-sigCh
+		c.Logger.Infof("received signal %v, shutting down gracefully", s)
+
+		if err := g.Shutdown(context.Background()); err != nil {
+			c.Logger.Errorf("error during grpc server shutdown: %v", err)
+		}
+	}()
+}
+
+func (g *grpcServer) Shutdown(ctx context.Context) error {
+	stopCh := make(chan struct{})
+	go func() {
+		g.server.GracefulStop()
+		close(stopCh)
+	}()
+
+	select {
+	case <-ctx.Done():
```

Comment:
So basically, for the Closers that do not require a context, should we wait using the `stopChan`, only to return when the close process is finished?

Do we need to make sure that the close does not close the client connection before any ongoing request/process is finished?

Reply:
I'm unsure what you meant. So, please let me rephrase.

We have 

When a signal is caught,
We create a context that expires with a timeout
Then we call, App.Shutdown with this context.

So the code in the Shutdown is supposed to stop all services, before the context expiration.

So every single closer we call should consider the context, even the one that don't have a context.

Which mean we need to exit the closer when the context is expired, otherwise we don't respect the timeout.

All this implies that we do are working with an app that is shutting down.

We are trying to gracefully stop the open connections before the timeout

It implies that yes, we might close opened connection if the clients/closer doesn't reply. But the app is stopped, so any new connection is refused.

So for me, we should get out each closer as soon the timeout is reached.

Is that clearer ?

That said, right now we are shutting down resources sequentially, we could shut them down in parallel, with a small refactoring. Of course, we might have to consider the consequence of doing that.

Reply:
> So maybe, we could copy this code in a function that accepts a function as argument in App.Shutdown\n\nSo you could use it for the following clients Kafka, Google, SQL …\n\nOr you copy paste this code in each Shutdowner/Closer called by App.Shutdown

This is the point where I would like to get feedbacks from you guys

Reply:
> It implies that yes, we might close opened connection if the clients/closer doesn't reply. But the app is stopped, so any new connection is refused.
> So for me, we should get out each closer as soon the timeout is reached.

Yes, we should not accept any new connection, also we need to finish the ongoing requests first.

> This is the point where I would like to get feedbacks from you guys

Yes we should have all the closers to use context for their closing, I think we can have a central closer that takes